docs/snapshot-controller.md

```diff
@@ -0,0 +1,132 @@
+<div align="center"><h1>Snapshot Controller</h1></div>
+
+```mermaid
+%%{init: {'theme':'forest'}}%%
+flowchart TD
+  %% Defining the styles
+    classDef Red fill:#FF9999;
+    classDef Amber fill:#FFDEAD;
+    classDef Green fill:#BDFFA4;
+
+  predicate((PREDICATE: <br>Snapshot got created OR <br> changed to Finished))
+
+  %%%%%%%%%%%%%%%%%%%%%%% Drawing EnsureAllIntegrationTestPipelinesExist() function 
+
+  %% Node definitions
+  ensure1(Process further if: Snapshot testing <br>is not finished yet)
+  areThereAnyITS{"Are there any <br>IntegrationTestScenario <br>present for the given <br>Application?"}
```

Comment:
That's a really good question, thanks for asking that.

You are right, and if I'm not wrong, you're referring to these 3 error handling cases present in the code ([first](https://github.com/redhat-appstudio/integration-service/blob/8981f8f01edd38b07070177d4f21ac70a2bb832a/controllers/snapshot/snapshot_adapter.go#L95), [second](https://github.com/redhat-appstudio/integration-service/blob/8981f8f01edd38b07070177d4f21ac70a2bb832a/controllers/snapshot/snapshot_adapter.go#L108), and [third](https://github.com/redhat-appstudio/integration-service/blob/8981f8f01edd38b07070177d4f21ac70a2bb832a/controllers/snapshot/snapshot_adapter.go#L117)). 





Personally, I do not want the diagrams to capture exactly how the code is written (I think it should be the other way around), because:
* then the diagram becomes tedious to read and wouldn't be a good quick-mind-map resource for us (Integration service engineers) -- and this is how I imagine us using these diagrams, as mind-maps.
* then the diagram would contain too much of implementation details that wouldn't matter to the general public (people from Product mgmt, engineers from other teams) who are just here to see the actual steps the controller performs, rather than seeing the error handlings unless it does something important.

so there's no rule as such to determine if we should include a specific error handling in the diagram or not, please feel free to experiment however you like! How I see it is to think if we're changing any resource's field/status in case of any errors, if yes, then that should be in the diagram. It would help us to answer questions like "Hey, which are places where we mark the Snapshot as invalid?".

Having said that, if I include the (_"second" and "third"_ -- linked above) error handling in the above block, it would look something like this:

```mermaid
%%{init: {'theme':'forest'}}%%
flowchart TD
  %% Defining the styles
    classDef Red fill:#FF9999;
    classDef Amber fill:#FFDEAD;
    classDef Green fill:#BDFFA4;
  %%%%%%%%%%%%%%%%%%%%%%% Drawing EnsureAllIntegrationTestPipelinesExist() function 

  %% Node definitions
  doesITSHasEnvDefined{Does the <br>IntegrationTestScenario <br>has any environment <br>defined in it?}
  skipCreatingTestPLR(Skip creating Test PLR for this ITS,<br> as it will be created by binding controller)
  createNewTestPLR(<b>Create a new Test PipelineRun</b> for each <br>of the above ITS, if it doesn't exists already)
  encounteredError1{"Encountered <br>error?"}
  requeueErrorOrStop("Call RequeueOnErrorOrStop()")
  markSnapshotInProgress(<b>Mark</b> Snapshot's Integration-testing <br>status as 'InProgress')
  fetchAllRequiredITS("Fetch all the required <br>(non-optional) IntegrationTestScenario <br>for the given Application")
  encounteredError2{"Encountered <br>error?"}
  logError("Log error")

  %% Node connections
  doesITSHasEnvDefined   --Yes--> skipCreatingTestPLR
  doesITSHasEnvDefined   --No-->  createNewTestPLR
  skipCreatingTestPLR    -->      fetchAllRequiredITS
  createNewTestPLR       -->      encounteredError1
  encounteredError1 --No-->  markSnapshotInProgress
  encounteredError1 --Yes--> requeueErrorOrStop
  markSnapshotInProgress --> encounteredError2
  encounteredError2 --Yes-->  logError
  encounteredError2 --No-->   fetchAllRequiredITS
  logError -->      fetchAllRequiredITS

  %% Assigning styles to nodes
  class encounteredError1,encounteredError2,logError,requeueErrorOrStop Red;

```

and I think this new addition doesn't really provide any new information.

Having said that, I'm open to knowing what you (and/or others) think about this, we can change this approach if we can agree on something better.

Reply:
@dheerajodha I agree that we don't need to draw error handling that doesn't change the state of resources or is a noticeable part of the logic of the adapter.

Reply:
This all makes sense, thanks for the clarification